datajoint/schema.py

```diff
@@ -1,15 +1,59 @@
+from operator import itemgetter
 import pymysql
 import logging
 
-from . import conn, DataJointError, NoDefinitionError
+from . import conn, DataJointError
+from datajoint.utils import to_camel_case
 from .heading import Heading
 from .base_relation import BaseRelation
-from .user_relations import Part
+from .user_relations import Part, Computed, Imported, Manual, Lookup
 import inspect
+
 logger = logging.getLogger(__name__)
 from warnings import warn
 
 
+class SchemaFactory:
+    """
+    Creates the appropriate python schema class objects from tables in a database.
+
+    SchemaFactory stores the class objects in a class dictionary and returns those
+    when prompted for the same table from the same database again. This way, the id
+    of both returned class objects is the same and comparison with python's "is" works
+    correctly.
+    """
+
+    table2class = {}
+
+    def __call__(self, database, table_name):
+        class_name = to_camel_case(table_name)
+
+        def _make_tuples(other, key):
+            raise NotImplementedError("This is an automatically created class. _make_tuples is not implemented.")
+
+        if (database, table_name) in SchemaFactory.table2class:
+            class_name, class_obj = SchemaFactory.table2class[database, table_name]
+        else:
+            if table_name.rfind('__') > 1:
+                master_name, master_class = self(database, table_name[:table_name.rfind('__')])
+                class_name = class_name[len(master_name):]
+                class_obj = type(class_name, (Part,), dict(definition=...))
+                setattr(master_class, class_name, class_obj)
+                class_name, class_obj = master_name, master_class
+
+            elif table_name.startswith('__'):
+                class_obj = type(class_name, (Computed,), dict(definition=..., _make_tuples=_make_tuples))
+            elif table_name.startswith('_'):
+                class_obj = type(class_name, (Imported,), dict(definition=..., _make_tuples=_make_tuples))
+            elif table_name.startswith('#'):
+                class_obj = type(class_name, (Lookup,), dict(definition=...))
+            else:
+                class_obj = type(class_name, (Manual,), dict(definition=...))
```

Comment:
rather than only looking at the prefixes, we should check the entire name by a regular expression.  For example, the current definition would interpret the job table `~jobs` as a manual table.